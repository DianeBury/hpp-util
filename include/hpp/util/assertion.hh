// Copyright (C) 2010 by Thomas Moulard, CNRS.
//
// This file is part of the hpp-util.
//
// This software is provided "as is" without warranty of any kind,
// either expressed or implied, including but not limited to the
// implied warranties of fitness for a particular purpose.
//
// See the COPYING file for more information.

#ifndef HPP_UTIL_ASSERTION_HH
# define HPP_UTIL_ASSERTION_HH
# include <boost/scope_exit.hpp>

# include <hpp/util/config.hh>
# include <hpp/util/exception.hh>

// If debug mode is disabled and assertions are not already
// disabled, disable them automatically.
# if (defined HPP_DEBUG) && (!defined HPP_ENABLE_ASSERTIONS)
#  define HPP_ENABLE_ASSERTIONS
# endif // (!defined HPP_DEBUG) && (!defined HPP_ENABLE_ASSERTIONS)

namespace hpp
{
  HPP_MAKE_EXCEPTION (HPP_UTIL_DLLAPI, AssertionError);
} // end of namespace hpp.

/// \brief Define HPP_ASSERT.
///
/// Throw an ::hpp::AssertionError if macro argument evaluates to
/// false.
# ifdef HPP_ENABLE_ASSERTIONS
#  define HPP_ASSERT(CONDITION)					\
  do {								\
    bool _x = (CONDITION);					\
    if (!_x)							\
      HPP_THROW_EXCEPTION					\
	(::hpp::AssertionError,					\
	 #CONDITION " evaluates to false");			\
  } while (0)
# else
#  define HPP_ASSERT(CONDITION)
# endif // HPP_ENABLE_ASSERTIONS

/// \brief Define macro for precondition checking.
# define HPP_PRECONDITION(CONDITION) HPP_ASSERT (CONDITION)

/// \brief Define macro for postcondition checking.
///
/// This macro allows an assertion to be done whenever a
/// scope is exited.
///
/// This macro takes two arguments: the condition which should
/// be evaluated and a list of captured variables.
/// All variables used in the condition must be in this list.
///
/// They can either be copied or taken by reference:
/// HPP_POSTCONDITION (a == 3, (a))
/// ...checks that a is equal to three and copy the variable a
/// when HPP_POSTCONDITION is called.
///
/// HPP_POSTCONDITION (a == 3, (&a))
/// ...takes a by reference and evaluated its final value (when
/// the scope is exited).
///
/// See Boost.ScopeExit for more details.
# define HPP_POSTCONDITION(CONDITION, CAPTURED_VARIABLES)	\
  BOOST_SCOPE_EXIT(CAPTURED_VARIABLES)				\
  {								\
    HPP_ASSERT (CONDITION);					\
  } BOOST_SCOPE_EXIT_END					\
  struct _e_n_d__w_i_t_h__s_e_m_i_c_o_l_o_n



#endif //! HPP_UTIL_ASSERTION_HH
