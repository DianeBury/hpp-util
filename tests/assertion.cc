// Copyright (C) 2010 by Thomas Moulard.
//
// This file is part of the hpp-util.
//
// hpp-util is free software: you can redistribute it and/or modify it
// under the terms of the GNU Lesser General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// hpp-util is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with hpp-util.  If not, see <http://www.gnu.org/licenses/>.

// Make sure assertions are enabled (otherwise this test does not make
// sense).
#ifndef HPP_ENABLE_ASSERTIONS
# define HPP_ENABLE_ASSERTIONS
#endif // !HPP_ENABLE_ASSERTIONS

#include "config.h"

#include <cassert>
#include <iostream>
#include <hpp/util/assertion.hh>

#include "common.hh"

int run_test ();
int my_plus_function (int a, int b);
int my_broken_plus_function (int a, int b);

int my_plus_function (int a, int b)
{
  // This algorithm does not work with negative numbers.
  HPP_PRECONDITION (a >= 0);
  HPP_PRECONDITION (b >= 0);

  int res = 0, expected = a + b;

  res = b;
  while (a > 0)
    --a, ++res;

  // This algorithm should have computed a + b.
  HPP_ASSERT (res == expected);

  return res;
}

int my_broken_plus_function (int a, int b)
{
  // This algorithm does not work with negative numbers.
  HPP_PRECONDITION (a >= 0);
  HPP_PRECONDITION (b >= 0);

  int res = 0, expected = a + b;

  res = b; // This is wrong.

  // This algorithm should have computed a + b.
  HPP_ASSERT (res == expected);

  return res;
}

int run_test ()
{
  // Assertion.
  try
    {
      int i = 4;
      HPP_ASSERT (i > 2);
      HPP_ASSERT (i < 3);
    }
  catch (::hpp::AssertionError& assertionError)
    {
      std::cout << assertionError << std::endl;
    }

  // Good calls.
  my_plus_function (4, 5);
  my_plus_function (7, 9);

  // Check for precondition failure.
  try
    {
      my_plus_function (-2, 5);
    }
  catch (::hpp::AssertionError& assertionError)
    {
      std::cout << assertionError << std::endl;
    }

  // Check for postcondition failure.
  try
    {
      my_broken_plus_function (3, 5);
    }
  catch (::hpp::AssertionError& assertionError)
    {
      std::cout << assertionError << std::endl;
    }

  return 0;
}

GENERATE_TEST ()
