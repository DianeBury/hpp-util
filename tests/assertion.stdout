../../tests/assertion.cc:75: i < 3 evaluates to false
../../tests/assertion.cc:39: a >= 0 evaluates to false
../../tests/assertion.cc:62: res == a + b evaluates to false
